### PR TITLE
Add additional basal stress parameters to the dynamics namelist

### PR DESCRIPTION
**Pull request by daveh150**
Added the following to dynamics_nml: k2, alphab, threshold_hw

For detailed information about submitting Pull Requests (PRs) to the CICE-Consortium,
please refer to: <https://github.com/CICE-Consortium/About-Us/wiki/Resource-Index#information-for-developers>

## PR checklist
- [x] Short (1 sentence) summary of your PR: 
    Moved basalstress parameters k2, alphab, threshold_hw as namelist parameters for ease in testing/changing.
- [x] Developer(s): 
    David Hebert, Rick Allard
- [x] Suggest PR reviewers from list in the column to the right.
- [x] Please copy the PR test results link or provide a summary of testing completed below.
[results.log](https://github.com/CICE-Consortium/CICE/files/4214372/results.log)

I also think this may be output from my report_results.csh:
https://github.com/CICE-Consortium/Test-Results/wiki/1561e76774.gordon.intel.20-02-16.135917

- How much do the PR code changes differ from the unmodified code? 
    - [x] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [x] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [x] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [ ] Yes
    - [x] No, does the documentation need to be updated at a later time?
        - [x] Yes - only if there is specific documentation for basalstress parameterization.
        - [ ] No 
- [ ] Please provide any additional information or relevant details below:

@eclare108213 edited this entry to fix the check-boxes (remove spaces).

### CODE REVIEW

**Comment by apcraig** [comment]
Waiting for a couple other reviews, but this looks good to me.

**Comment by daveh150** [comment]
Thank you Tony. We have been working on landfast ice. The thought was to have the basalstress parameters in a namelist for easy manipulation. Eventually we are working toward a 2D k2 parameter that varies over regions, but for now we have been testing with a variable k2 parameter.

 

Dave

 

David A. Hebert

Ocean Sciences Division, Code 7322 
Naval Research Laboratory 
Stennis Space Center, MS 39529
(228)688-5846; DSN 828-5846 
http://www.nrl.navy.mil <http://www.nrl.navy.mil/>

**Comment by daveh150** [comment]
Thank you Philippe, I changed the title to be as you suggested.

 

David

 

David A. Hebert

Ocean Sciences Division, Code 7322 
Naval Research Laboratory 
Stennis Space Center, MS 39529
(228)688-5846; DSN 828-5846 
http://www.nrl.navy.mil <http://www.nrl.navy.mil/>

**Comment by apcraig** [comment]
@daveh150, we are going to probably freeze the code for the next minor release tomorrow.  This PR is not critical for the release, but it would be great to include.  I think the only thing that's needed is the updated documentation.  If you are able to take care of that this week, that's great.  If not, that's fine too, we'll just merge this PR later when it is done.  If you need any help with the documentation, let me know.  Thanks.

**Comment by daveh150** [comment]
I just committed documentation updates for the movement of basalstress parameters to namelist. The updates were minor, mostly needed to updated the info in ug_case_settings.rst. Please let me know if these need to be further changed/elaborated.  Thank you!

File: doc/source/user_guide/ug_case_settings.rst [hunk 1]
```diff
@@ -275,6 +275,9 @@ Table of namelist options
    "","", "``-1``", "Transport Disabled", ""
    "","``basalstress``", "true/false", "use basal stress parameterization for landfast ice", ""
    "","``k1``", "real", "1st free parameter for landfast parameterization", "8."
+   "","``k2``", "real", "2nd free parameter (N/m^3) for landfast parameterization", "15."
+   "","``alphab``", "real", ":math:`\alpha_{b}` factor in :cice:`Lemieux16`", "20."
```
Comment (apcraig) [inline doc/source/user_guide/ug_case_settings.rst]: The Lemieux16 is not being resolved properly in the documentation.  See https://cice-consortium-cice--404.org.readthedocs.build/en/404/user_guide/ug_case_settings.html#table-of-namelist-options.  Could this be corrected?

**Comment by daveh150** [comment]
Is suspect it is because it is a type-o (my type-o, BTW). Looks like my fingers are conditioned to  type ‘cice’ instead of ‘cite’. (J) 

 

i.e., It think it should be 

 

:cite:`Lemieux16`

 

I just pushed a correction to my GITHUB repo. Please let me know if I need to do anything else to fix this.

 

Thanks!

Dave

 

David A. Hebert

Ocean Sciences Division, Code 7322 
Naval Research Laboratory 
Stennis Space Center, MS 39529
(228)688-5846; DSN 828-5846 
http://www.nrl.navy.mil <http://www.nrl.navy.mil/>

**Comment by daveh150** [comment]
Question: I have a separate item I’d like to add for the next release. It is a minor change to ice_open. I do not know how to make a separate pull request, it seems to go back to this one. Is it possible to make a separate pull request when one is open already? Or should I wait until this one is cleared?

 

BTW, the change is related to integer overflows for large grids. Our Global grid is 9000x7054. When opening our grid file in 64 bit mode, we get an overflow in computing the RecSize: 9000x7054x64 is larger than signed 4byte integer can handle.

 

Thanks,
Dave

 

David A. Hebert

Ocean Sciences Division, Code 7322 
Naval Research Laboratory 
Stennis Space Center, MS 39529
(228)688-5846; DSN 828-5846 
http://www.nrl.navy.mil <http://www.nrl.navy.mil/>

**Comment by apcraig** [comment]
The documentation error seems to be fixed.  You can check the readthedocs output by clicking on the readthedocs link in the "checks" section above.  Click on "show all checks", then click on "details" on the readthedocs line, and you'll be taken to the documentation for this pull request where you can review the sections you have modified.

@daveh150, In general, if you want to create a new pull request you have to create a new branch.  It looks like you are current working in your master branch which is NOT what we recommend.  We recommend you create branches for each feature and that master be synced up with the consortium master regularly.  This is documented in the git workflow guide, https://github.com/CICE-Consortium/About-Us/wiki/Git-Workflow-Guidance.

Another way to deal with the large integer issue is to change the variable to an integer*8.  It seems even with the current fix, we are less than a factor of 10 from blowing out the integer*4 even with this fix.  I'm OK with the change as implemented, but we'll probably need to update it again at some point.

File: cicecore/cicedynB/infrastructure/ice_read_write.F90 [hunk 2]
```diff
@@ -99,10 +102,15 @@ subroutine ice_open(nu, filename, nbits, algn)
             open(nu,file=filename,form='unformatted')
 
          else                   ! direct access
-            RecSize = nx_global*ny_global*nbits/8
+            !RecSize = nx_global*ny_global*nbits/8  ! dah: original
```
Comment (phil-blain) [inline cicecore/cicedynB/infrastructure/ice_read_write.F90]: I think we should define a variable for this `8` magic number.

**Comment by daveh150** [comment]
Thank you Tony, I’ll make new branches for code updates in the future, sorry for not realizing that earlier. I reverted my master branch to just the basalstress changes, and will make new branch for the integer overflow issue.  

 

For the integer overflow issue, nx_global and ny_global are defined as 4 byte integers. For your thought of integer8, if we do not change the grid variable size, I believe we would need to define the nbits variable as an 8 byte integer to promote the integer computation to 8 bytes, as defining RecSize alone as integer8 will not work in this case. I did not try defining nbits as 8 byte integer, I instead used this fix that Alan Wallcraft implemented from our GOFS code base. 

 

I can define nbits as integer8, and define the 8 as a parameter from icepack (in a new branch!),  if that will help.

 

Thanks,
Dave

 

David A. Hebert

Ocean Sciences Division, Code 7322 
Naval Research Laboratory 
Stennis Space Center, MS 39529
(228)688-5846; DSN 828-5846 
http://www.nrl.navy.mil <http://www.nrl.navy.mil/>

**Comment by apcraig** [comment]
Thanks @daveh150.  You are right we might have to play some tricks with over variables to get the integer*8 to work properly, but it should not be hard.  I'm also fine with the implementation you proposed and included in this PR.  For smaller changes, I think it's OK to group them together a little.   If you want to include it again, I think that's fine.

The readthedocs check seems to be stuck and I can't retrigger it manually.  Could you do a 
```
  git commit -m "retrigger checks" --allow-empty
  git push origin master
```
Or if you want to put the integer overflow back into the PR, that should trigger it as well.  Thanks!

**Comment by daveh150** [comment]
Hi Tony,

 

I ran the git commands you listed below. Let me known if that doesn’t work.

 

I am almost ready to make a PR for the code change, defining nbits as an 8 byte integer.  Hopefully by the end of the day today.

 

Thank you,

Dave

 

David A. Hebert

Ocean Sciences Division, Code 7322 
Naval Research Laboratory 
Stennis Space Center, MS 39529
(228)688-5846; DSN 828-5846 
http://www.nrl.navy.mil <http://www.nrl.navy.mil/>

**Comment by apcraig** [comment]
All the tests are passing now.  I don't think there are any outstanding issues.  I will merge.

**Pull request merged**